vapi/simulator/simulator.go

```diff
@@ -967,19 +961,309 @@ func (s *handler) library(w http.ResponseWriter, r *http.Request) {
+	// a subscribed library item.
+	if dstLib == nil {
+		var ok bool
+		if dstLib, ok = s.Library[dstItem.LibraryID]; !ok {
+			return fmt.Errorf("cannot find sub library id %q", dstItem.LibraryID)
+		}
+	}
+
+	// srcLib is nil when this function is used to synchronize an individual
+	// item versus synchronizing the entire library.
+	if srcLib == nil {
+		sub := dstLib.Subscription
+		if sub == nil {
+			return nil
+		}
+		var ok bool
+		srcLibID := path.Base(sub.SubscriptionURL)
+		if srcLib, ok = s.Library[srcLibID]; !ok {
+			return fmt.Errorf("cannot find pub library id %q", srcLibID)
+		}
+	}
+
+	// Get the path to the destination library item on the local filesystem.
+	dstItemPath := libraryPath(dstLib.Library, dstItem.ID)
+
+	// Get the source item.
+	srcItem, ok := srcLib.Item[dstItem.SourceID]
+	if !ok {
+		// The source item is no more, so delete the destination item.
+		delete(dstLib.Item, dstItem.ID)
+
+		// Clean up the destination item's files as well.
+		os.RemoveAll(dstItemPath)
+
+		return nil
+	}
+
+	// There is nothing to sync if the content version has not changed, the
+	// item is already cached, and syncAll is false.
+	if dstItem.ContentVersion == srcItem.ContentVersion &&
+		dstItem.Cached && !syncAll {
+
+		return nil
+	}
+
+	// lastSyncTime is nil when this function is used to synchronize an
+	// individual item versus synchronizing the entire library.
+	if lastSyncTime == nil {
+		now := time.Now().UTC()
+		lastSyncTime = &now
+	}
+	dstItem.LastSyncTime = lastSyncTime
+
+	// Copy information from the srcItem to dstItem.
+	dstItem.Name = srcItem.Name
+	dstItem.ContentVersion = srcItem.ContentVersion
+	dstItem.Type = srcItem.Type
+	dstItem.Description = srcItem.Description
+	dstItem.Version = srcItem.Version
+
+	// Update the destination item's files from the source.
+	dstItem.File = make([]library.File, len(srcItem.File))
+	copy(dstItem.File, srcItem.File)
+
+	// If the destination item was previously cached or syncAll was used, then
+	// mark the destination item as cached.
+	dstItem.Cached = dstItem.Cached || syncAll
+	fileIsCached := true
+	fileIsNotCached := false
+	fileZeroSize := int64(0)
+
+	// Ensure a directory exists on the local filesystem for the destination
+	// item.
+	if err := os.MkdirAll(dstItemPath, 0750); err != nil {
+		return fmt.Errorf(
+			"failed to make directory for library %q item %q: %w",
+			dstLib.ID,
+			dstItem.ID,
+			err)
+	}
+
+	// Update the the destination item's files.
+	srcItemPath := libraryPath(srcLib.Library, srcItem.ID)
+	for i := range dstItem.File {
+		var (
+			dstFile = &dstItem.File[i]
+			srcFile = srcItem.File[i]
+		)
+
+		if !isValidFileName(dstFile.Name) || !isValidFileName(srcFile.Name) {
+			return errors.New("invalid file name")
+		}
+
+		var (
+			dstFilePath = path.Join(dstItemPath, dstFile.Name)
+			srcFilePath = path.Join(srcItemPath, srcFile.Name)
+		)
+
+		// .ovf and .mf files are always cached.
+		if ovfOrManifestRx.MatchString(dstFile.Name) {
```

Comment:
nit: CL doesn't store the manifest file in sub lib.

Reply:
Ack. I will remove it then, thanks!